### PR TITLE
(PPS-635): chore(security): update urllib3 and related deps

### PR DESCRIPTION
**Pull request by Avantol13**
Jira Tickets: 

- [PPS-635](https://ctds-planx.atlassian.net/browse/PPS-635)
- [PPS-656](https://ctds-planx.atlassian.net/browse/PPS-656)
- [PPS-657](https://ctds-planx.atlassian.net/browse/PPS-657)

Relies on:

- https://github.com/uc-cdis/cirrus/pull/94
- https://github.com/uc-cdis/cdisutils-test/pull/31

### Improvements
- dependency cleanup. storageclient code migrated from multiple libraries (causing circular dependency issues) to Fence, the only place where the code was used

### Dependency updates
- update urllib3

### Deployment changes

[PPS-635]: https://ctds-planx.atlassian.net/browse/PPS-635?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[PPS-656]: https://ctds-planx.atlassian.net/browse/PPS-656?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[PPS-657]: https://ctds-planx.atlassian.net/browse/PPS-657?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by k-burt-uch — APPROVED** [review]
Checked this branch out locally in a fresh venv, ran poetry install without issue. pytest ran without issue (just confirming things GH actions already told us hahaha) 

Approved.

**Review by k-burt-uch — APPROVED** [review]

File: fence/resources/storage/storageclient/cleversafe.py [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+                "List buckets failed with code: {code}".format(
+                    code=response.status_code
+                )
+            )
+            raise RequestError(response.text, response.status_code)
+
+    def create_user(self, name):
+        """
+        Creates a user
+        TODO Input sanitazion for parameters
```
Comment (k-burt-uch) [inline fence/resources/storage/storageclient/cleversafe.py:272]: Do we have a ticket for this? ?

File: fence/resources/storage/storageclient/cleversafe.py [hunk 2]
```diff
@@ -0,0 +1,531 @@
...
+        Get an acl object and add the missing credentials
+        to the one retrieved from the target bucket
+        new_grants contains a list of users and permissions
+        [('user1', ['read-storage', 'write-storage']),...]
+        """
+        user_id_list = []
+        for user in new_grants:
+            user_id_list.append(self._get_user_id(user[0]))
+        bucket_id = self._get_bucket_id(bucket)
+        response = self._get_bucket_by_id(bucket_id)
```
Comment (k-burt-uch) [inline fence/resources/storage/storageclient/cleversafe.py:456]: Is this bucket -> bucket_id -> bucket is just a check that the bucket information we have is up to date with whatever this "viewSystem.adm" endpoint responds with?

File: fence/resources/storage/storageclient/cleversafe.py [hunk 3]
```diff
@@ -0,0 +1,531 @@
...
+    def add_bucket_acl(self, bucket, username, access=[]):
+        """
+        Add permissions to a user on the bucket ACL
+        """
+        try:
+            bucket_param = "vaultUserPermissions[{0}]".format(
+                self._get_bucket_id(bucket)
+            )
+        except KeyError:
+            msg = "Bucket {0} wasn't found on the database"
```
Comment (k-burt-uch) [inline fence/resources/storage/storageclient/cleversafe.py:492]:
```suggestion
            msg = "Bucket {0} wasn't found in the database"
```

File: fence/resources/storage/storageclient/cleversafe.py [hunk 4]
```diff
@@ -0,0 +1,531 @@
...
+        try:
+            access_lvl = max(self._permissions_order[role] for role in access)
+            data = {
+                "id": self._get_user_id(username),
+                bucket_param: self._permissions_value[access_lvl],
+            }
+            if access_lvl == 3:
+                data["rolesMap[vaultProvisioner]"] = "true"
+        except KeyError:
+            msg = "User {0} wasn't found on the database"
```
Comment (k-burt-uch) [inline fence/resources/storage/storageclient/cleversafe.py:504]:
```suggestion
            msg = "User {0} wasn't found in the database"
```

File: fence/resources/storage/storageclient/cleversafe.py [hunk 5]
```diff
@@ -0,0 +1,531 @@
...
+            self.logger.error(msg.format(username))
+            raise NotFoundError(msg.format(username))
+        response = self._request("POST", "editAccount.adm", payload=data)
+        if response.status_code != 200:
+            msg = "Error trying to change buket permissions for user {0}"
```
Comment (k-burt-uch) [inline fence/resources/storage/storageclient/cleversafe.py:509]:
```suggestion
            msg = "Error trying to change bucket permissions for user {0}"
```

File: fence/resources/storage/storageclient/cleversafe.py [hunk 6]
```diff
@@ -0,0 +1,531 @@
...
+        else:
+            self.logger.error(
+                "Delete keypair failed with code: {0}".format(response.status_code)
+            )
+            raise RequestError(response.text, response.status_code)
+
+    def delete_all_keypairs(self, name):
+        """
+        Remove all keys from a give user
+        TODO Make this robust against possible errors so most of the keys are deleted
```
Comment (k-burt-uch) [inline fence/resources/storage/storageclient/cleversafe.py:322]: Jira ticket

File: fence/resources/storage/storageclient/cleversafe.py [hunk 7]
```diff
@@ -0,0 +1,531 @@
...
+            host=self._host, oper=operation
+        )
+        url = base_url + "?" + urlencode(dict(**kwargs))
+        return requests.request(
+            method,
+            url,
+            auth=self._auth,
+            data=payload,
+            verify=config["VERIFY_CLEVERSAFE_CERT"],
+        )  # self-signed certificate
```
Comment (k-burt-uch) [inline fence/resources/storage/storageclient/cleversafe.py:192]: ? Is this comment meant to indicate that verify should be false if using a self-signed cert?

File: fence/resources/storage/storageclient/google.py [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+
+        Google architecture requires that a separate process populate
+        a proxy Google group per user. If it doesn't exist, we can't create it
+        here.
+        """
+        user_proxy = self.get_user(username)
+        if not user_proxy:
+            raise Exception(
+                "Unable to determine User's Google Proxy group. Cannot create "
+                "here. Another process should create proxy groups for "
```
Comment (k-burt-uch) [inline fence/resources/storage/storageclient/google.py:85]: Can we point them to the process used or is it so case-by-case there's not really a way to do it? Trying to head off a user-services/dev-oncall question if someone sees this.

File: fence/resources/storage/storageclient/google.py [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+
+    def create_keypair(self, username):
+        """
+        Creates a keypair for the user, and
+        returns it
+        """
+        msg = "create_keypair not implemented"
+        raise NotImplementedError(msg)
+
+    def delete_keypair(self, username, access_key):
```
Comment (k-burt-uch) [inline fence/resources/storage/storageclient/google.py:99]: Are these only cleversafe specific implementations? If so, why are they in the ABC?? There's a lot of unimplemented methods in this implementation of storage client.

File: fence/resources/storage/storageclient/google.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+            access (str): IGNORED. For Google buckets, the Google Bucket Access
+                Group is given access to the bucket through Google's
+                IAM, so you cannot selectively choose permissions. Once you're
+                added, you have the access that was set up for that group
+                in Google IAM.
+        """
+        response = None
+        with GoogleCloudManager(project_id=self.google_project_id) as g_mgr:
+            try:
+                response = g_mgr.add_member_to_group(
```
Comment (k-burt-uch) [inline fence/resources/storage/storageclient/google.py:127]: Just a double check:
https://github.com/uc-cdis/cirrus/blob/be8631dccf063e3bf4ce702fe81daaf8253fc045/cirrus/google_cloud/manager.py#L1304 

Does this handle API rate limits we might hit so we don't have to worry about that here??

**Review by k-burt-uch — CHANGES_REQUESTED** [review]
I have some concerns about the storage client abstraction. Given this is a security issue wiht a limited timeframe, its not a show stopper. 

However, TODOs should be associated with JIRA tickets, especially ones related to input sanitization.

File: fence/resources/storage/storageclient/cleversafe.py (hunk 1 above)
Comment (Avantol13) [reply fence/resources/storage/storageclient/cleversafe.py:272]: All this code is copy pasted verbatim from the existing client library

So there's an argument that all this have been here for years. 

I can go through and clean things up, but honestly I avoided touching this at all to preserve a clean history of an exact copy paste from the library.

Let me know what you think

File: fence/resources/storage/storageclient/google.py (hunk 10 above)
Comment (Avantol13) [reply fence/resources/storage/storageclient/google.py:127]: It has backoff setup, yeah

File: fence/resources/storage/storageclient/cleversafe.py (hunk 1 above)
Comment (k-burt-uch) [reply fence/resources/storage/storageclient/cleversafe.py:272]: That's fair. I'll go through and add tickets to the TODO in a separate PR.

**Comment by Avantol13** [comment]
https://ctds-planx.atlassian.net/browse/PXP-11238

Ticket to do storageclient cleanup in the future on a separate PR

**Pull request merged**